opts/filters.go

```diff
@@ -0,0 +1,98 @@
+package dhtopts
+
+import (
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr-net"
+)
+
+// PublicQueryFilter returns true if the peer is suspected of being publicly accessible
+func PublicQueryFilter(h host.Host, ai peer.AddrInfo) bool {
+	if len(ai.Addrs) == 0 {
+		return false
+	}
+
+	var hasPublicAddr bool
+	for _, a := range ai.Addrs {
+		if isRelayAddr(a) {
+			return false
+		}
+		if manet.IsPublicAddr(a) {
+			hasPublicAddr = true
+		}
+	}
+	return hasPublicAddr
+}
+
+// PublicRoutingTableFilter allows a peer to be added to the routing table if the connections to that peer indicate
+// that it is on a public network
+func PublicRoutingTableFilter(conns []network.Conn) bool {
+	for _, c := range conns {
+		addr := c.RemoteMultiaddr()
+		if !isRelayAddr(addr) && manet.IsPublicAddr(addr) {
+			return true
+		}
+	}
+	return false
+}
+
+// PrivateQueryFilter returns true if the peer is suspected of being accessible over a shared private network
+func PrivateQueryFilter(h host.Host, ai peer.AddrInfo) bool {
+	conns := h.Network().ConnsToPeer(ai.ID)
+	if len(conns) > 0 {
+		for _, c := range conns {
+			if manet.IsPrivateAddr(c.RemoteMultiaddr()) {
+				return true
+			}
+		}
+		return false
+	}
+
+	if len(ai.Addrs) == 0 {
+		return false
+	}
+
+	var hasPrivateAddr bool
+	for _, a := range ai.Addrs {
+		if manet.IsPublicAddr(a) {
+			if !isRelayAddr(a) {
+				return false
```

Comment:
This check is perhaps overly aggressive because it's possible that a peer has both a private address and a public one. Perhaps for queries it's not worth doing this check and only being critical of routing table additions.

Reply:
What's the idea here? Why would we want to filter out peers that have public addresses? IIRC, peers on the public network should also be in the private DHT.

Reply:
We shouldn't filter out peers with public addresses. All peers on the local network, dialable or not, should be part of the local DHT.

Notes:

1. We should have some way to pass an address filter function down into the dialer in the future. That's probably the right way to handle this.
2. We should apply all the checks from https://github.com/libp2p/go-libp2p-kad-dht/pull/471/files#r387453724. E.g., we should filter out peers when (a) we know our public IPv4 address and (b) we see that it's _not_ the same as theirs.